frame/staking/src/lib.rs

```diff
@@ -2581,9 +2624,15 @@ impl<T: Trait> Module<T> {
 		}).collect::<Result<Vec<T::AccountId>, Error<T>>>()?;
 
 		// decode the rest of the snapshot.
-		let snapshot_nominators = <Module<T>>::snapshot_nominators()
+		let snapshot_nominators = Self::snapshot_nominators()
 			.ok_or(Error::<T>::SnapshotUnavailable)?;
 
+		// rest of the size of the solution must be correct.
+		ensure!(
+			snapshot_nominators.len() as u32 == election_size.nominators,
+			Error::<T>::PhragmenBogusElectionSize,
+		);
```

Comment:
Why don't you do this check before actually decoding the validators up above? You can use the same `decode_len()` technique, right?

Reply:
Fair point but I see it is a dilemma and I chose to do otherwise. The validator one is needed for a more important check that even non-malicious submissions could check: that you need to provide a correct number of winners. 

This is not the case here. For nominators I see doing `decode_len` a waste when you are already reading the value later.

Reply:
The "attack vector" I see is passing in a small number in `election_size` which would lead to small weight (and associated fees) and then forcing you to load and decode a lot of nominators because the size in storage might be quite big.
I don't think it's very practical, but the basic theoretical thing is there.

Reply:
Yes I also think it is better to do decode_len here, in case of malicious input we don't want to decode the entire nominator snapshot (which is big) if the user hasn't provided a correct value.

Reply:
There is no dilemma, btw. If you use `decode_len` you can just put a check before you start loading actual data from storage.

Reply:
Fair enough 👍  thanks for the hint.

Reply:
https://github.com/paritytech/substrate/pull/6032/commits/e9371fba84ce8aa5fe8796f254aa4d685e810735